darts/utils/optimised_historical_forecasts.py

```diff
@@ -0,0 +1,427 @@
+        )
+        freq = series_.freq
+
+        # obtain forecastable indexes boundaries, adjust target & covariates boundaries accordingly
+        (
+            hist_fct_start,
+            hist_fct_end,
+            hist_fct_tgt_start,
+            hist_fct_tgt_end,
+            hist_fct_pc_start,
+            hist_fct_pc_end,
+            hist_fct_fc_start,
+            hist_fct_fc_end,
+        ) = _get_historical_forecast_boundaries(
+            model=model,
+            series=series_,
+            series_idx=idx,
+            past_covariates=past_covariates_,
+            future_covariates=future_covariates_,
+            start=start,
+            forecast_horizon=forecast_horizon,
+            overlap_end=overlap_end,
+            freq=freq,
+            show_warnings=show_warnings,
+        )
+
+        # Additional shift, to account for the model output_chunk_length
+        shift_start = 0
+        shift_end = 0
+        if model.output_chunk_length > 1:
+            # used to convert the shift into the appropriate unit
+            unit = freq if series_.has_datetime_index else 1
+
+            if not model.multi_models:
+                shift_start = model.output_chunk_length - 1
+
+                hist_fct_tgt_start -= shift_start * unit
+                hist_fct_pc_start -= shift_start * unit
+                hist_fct_fc_start -= shift_start * unit
+
+            if model.output_chunk_length == forecast_horizon:
+                shift_end = model.output_chunk_length - 1
+
+                hist_fct_tgt_end += shift_end * unit
+                hist_fct_pc_end += shift_end * unit
+                hist_fct_fc_end += shift_end * unit
+
+        X, _ = create_lagged_prediction_data(
+            target_series=series_[hist_fct_tgt_start:hist_fct_tgt_end],
+            past_covariates=None
+            if past_covariates_ is None
+            else past_covariates_[hist_fct_pc_start:hist_fct_pc_end],
+            future_covariates=None
+            if future_covariates_ is None
+            else future_covariates_[hist_fct_fc_start:hist_fct_fc_end],
+            lags=model.lags.get("target", None),
+            lags_past_covariates=model.lags.get("past", None),
+            lags_future_covariates=model.lags.get("future", None),
+            uses_static_covariates=model.uses_static_covariates,
+            last_static_covariates_shape=model._static_covariates_shape,
+            max_samples_per_ts=None,
+            check_inputs=True,
+            use_moving_windows=True,
+            concatenate=False,
+        )
+
+        # stride must be applied post-hoc to avoid missing values
+        X = X[0][:, :, 0]
+
+        # repeat rows for probabilistic forecast
+        forecast = model._predict_and_sample(
+            np.repeat(X, num_samples, axis=0), num_samples
+        )
+
+        require_auto_regression: bool = forecast_horizon > model.output_chunk_length
```

Comment:
we can remove everything autoregressive related as we don't support it yet.
We can add this whenever we add support for it :)